### PR TITLE
Fix EM_ASM not working with setjmp/longjmp

### PR DESCRIPTION
**Pull request by quantum5**
Refs https://github.com/emscripten-core/emscripten/issues/8894. This fix does not handle dynamic linking, which requires additional work.

This is a fixed version of #2271.

### CODE REVIEW

**Comment by kripken** [comment]
What was wrong before / what is the fix?

**Comment by quantum5** [comment]
There were two bugs:

* Just because you can't get a constant index doesn't mean you should crash. If you pass a function pointer around, for example, `invoke_*` would have been called on a non-const function pointer. However, this is fine for `EM_ASM` because we are always calling `emscripten_asm_const_*` directly.
* Some locals are not set, for example, function arguments. In those cases, the old code had an infinite loop. This is fixed by bailing out of the loop if `local.set` cannot be found.

File: src/wasm/wasm-emscripten.cpp [hunk 1]
```diff
@@ -702,45 +717,110 @@ void AsmConstWalker::noteNonLinear(Expression* curr) {
...
+        assert(set->index == get->index);
+        arg = set->value;
+      } else {
+        reportError(arg);
+        return 0;
+      }
+    } else if (arg->is<GlobalGet>()) {
+      // In the dynamic linking case, indices start at __table_base.
+      // We want the value relative to __table_base.
+      return 0;
```
Comment (tlively) [inline src/wasm/wasm-emscripten.cpp]: It is unclear to me how returning 0 fulfills the requirements stated in the comment. Perhaps the comment can be clarified?

File: src/wasm/wasm-emscripten.cpp [hunk 2]
```diff
@@ -702,45 +717,110 @@ void AsmConstWalker::noteNonLinear(Expression* curr) {
...
+      Fatal() << "Unexpected arg0 type (" << getExpressionName(arg)
+              << ") in call to to: " << target;
+    }
+  }
+  return arg->cast<Const>();
+}
+
+Index AsmConstWalker::resolveConstIndex(
+  Expression* arg, std::function<void(Expression*)> reportError) {
+  while (!arg->dynCast<Const>()) {
```
Comment (tlively) [inline src/wasm/wasm-emscripten.cpp]: Can use `is` instead of `dynCast` here.

File: src/wasm/wasm-emscripten.cpp [hunk 3]
```diff
@@ -812,10 +894,27 @@ void AsmConstWalker::queueImport(Name importName, std::string baseSig) {
   queuedImports.push_back(std::unique_ptr<Function>(import));
 }
 
+Literal AsmConstWalker::tableIndexForName(Name name) {
+  if (tableIndices.count(name)) {
```
Comment (tlively) [inline src/wasm/wasm-emscripten.cpp]: Would it be better to use `find` here rather than `count`?

File: src/wasm/wasm-emscripten.cpp [hunk 4]
```diff
@@ -702,45 +717,110 @@ void AsmConstWalker::noteNonLinear(Expression* curr) {
...
-
-    auto* value = arg->cast<Const>();
+    auto* value = resolveConstAddr(curr->operands[0], import->base);
     auto code = codeForConstAddr(wasm, segmentOffsets, value);
     sigsForCode[code].insert(sig);
 
     // Replace the first argument to the call with a Const index
     Builder builder(wasm);
     curr->operands[0] = builder.makeConst(idLiteralForCode(code));
+  } else if (import->base.startsWith(INVOKE_PREFIX)) {
```
Comment (tlively) [inline src/wasm/wasm-emscripten.cpp]: This code dealing with `INVOKE_PREFIX`es seems to be new. Can you add a comment explaining what is happening here and why?

File: src/wasm/wasm-emscripten.cpp [hunk 5]
```diff
@@ -702,45 +717,115 @@ void AsmConstWalker::noteNonLinear(Expression* curr) {
...
+        assert(set->index == get->index);
+        arg = set->value;
+      }
+    } else if (auto* value = arg->dynCast<Binary>()) {
+      // In the dynamic linking case the address of the string constant
+      // is the result of adding its offset to __memory_base.
+      // In this case are only looking for the offset with the data segment so
+      // the RHS of the addition is just what we want.
+      assert(value->op == AddInt32);
+      arg = value->right;
```
Comment (tlively) [inline src/wasm/wasm-emscripten.cpp]: Will we ever run into a chain of expressions that contain more than one binary expression? I'm wondering if we should be even more strict about what we accept, perhaps with a flag saying the next expression must be the final constant.

Comment (quantum5) [reply src/wasm/wasm-emscripten.cpp]: I didn't write this code. It's the result of method extraction.

**Pull request merged**